source/tests/system/nifgen_driver_api_tests.cpp

```diff
@@ -285,6 +334,80 @@ class NiFgenDriverApiTest : public ::testing::Test {
     return response.status();
   }
 
+  void allocate_named_waveform(std::string channel_name, std::string waveform_name, int waveform_size)
+  {
+    configure_output_mode(channel_name.c_str(), fgen::OutputMode::OUTPUT_MODE_NIFGEN_VAL_OUTPUT_SCRIPT);
+    ::grpc::ClientContext context;
+    fgen::AllocateNamedWaveformRequest request;
+    request.mutable_vi()->set_id(GetSessionId());
+    request.set_channel_name(channel_name);
+    request.set_waveform_name(waveform_name);
+    request.set_waveform_size(waveform_size);
+    fgen::AllocateNamedWaveformResponse response;
+
+    ::grpc::Status status = GetStub()->AllocateNamedWaveform(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    expect_api_success(response.status());
+  }
+
+  void write_named_waveform_f64(std::string channel_name, std::string waveform_name, int waveform_size, double waveform[])
+  {
+    ::grpc::ClientContext context;
+    fgen::WriteNamedWaveformF64Request request;
+    request.mutable_vi()->set_id(GetSessionId());
+    request.set_channel_name(channel_name);
+    request.set_waveform_name(waveform_name);
+    request.mutable_data()->CopyFrom(google::protobuf::RepeatedField<double>(waveform, waveform + waveform_size));
+    fgen::WriteNamedWaveformF64Response response;
+
+    ::grpc::Status status = GetStub()->WriteNamedWaveformF64(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    expect_api_success(response.status());
+  }
+
+  ViInt32 allocate_waveform(std::string channel_name, int waveform_size)
+  {
+    configure_output_mode(channel_name.c_str(), fgen::OutputMode::OUTPUT_MODE_NIFGEN_VAL_OUTPUT_SEQ);
+    ::grpc::ClientContext context;
+    fgen::AllocateWaveformRequest request;
+    request.mutable_vi()->set_id(GetSessionId());
+    request.set_channel_name(channel_name);
+    request.set_waveform_size(waveform_size);
+    fgen::AllocateWaveformResponse response;
+
+    ::grpc::Status status = GetStub()->AllocateWaveform(&context, request, &response);
+
+    EXPECT_TRUE(status.ok());
+    expect_api_success(response.status());
+
+    return response.waveform_handle();
+  }
+
+  void write_waveform_complexf64(std::string channel_name, int waveform_size, int waveform_handle, NIComplexNumber waveform[])
+  {
+    set_bool_attribute(channel_name.c_str(), fgen::NiFgenAttributes::NIFGEN_ATTRIBUTE_OSP_ENABLED, true);
```

Comment:
These two `set*` calls are not really part of "WriteWaveformComplexF64". They might be required for calling this API, but are not part of making grpc call to the API. Keep calls `set_bool_attribute` and `set_int32_attribute` in the caller of `write_waveform_complexf64` itself.

Reply:
Same for calls to `configure_output_mode` in the two helpers above.